geomstats/symmetric_positive_definite_matrices.py

```diff
@@ -0,0 +1,216 @@
+    """
+    dim_vec = len(vector)
+    dim_mat = (np.sqrt(8 * dim_vec + 1) - 1) / 2
+    matrix = np.zeros((dim_mat, dim_mat))
+
+    idx = 0
+    for j in range(dim_mat):
+        for i in range(j + 1):
+            if i == j:
+                matrix[i, j] = vector[idx]
+            else:
+                matrix[i, j] = vector[idx] / np.sqrt(2)
+            idx += 1
+
+    for j in range(dim_mat):
+        for i in range(j + 1, dim_mat):
+            matrix[i, j] = matrix[j, i]
+
+    return matrix
+
+
+def riemannian_inner_product(ref_point, tangent_vec_a, tangent_vec_b):
+    """
+    Compute the inner product of tangent_vec_a and tangent_vec_b
+    at point ref_point using the affine invariant Riemannian metric.
+    """
+    inv_ref_point = np.linalg.inv(ref_point)
+
+    aux_a = np.dot(inv_ref_point, tangent_vec_a)
+    aux_b = np.dot(inv_ref_point, tangent_vec_b)
+
+    inner_product = np.trace(np.dot(aux_a, aux_b))
+
+    return inner_product
+
+
+def riemannian_exp(ref_point, tangent_vec):
+    """
+    Compute the Riemannian exponential at point ref_point
+    of tangent vector tangent_vec wrt the metric
+    defined in riemannian_inner_product.
+
+    This gives a symmetric positive definite matrix.
+    """
+    sqrt_ref_point = scipy.linalg.sqrtm(ref_point)
+    inv_sqrt_ref_point = np.linalg.inv(sqrt_ref_point)
+
+    tangent_vec_at_id = np.dot(np.dot(inv_sqrt_ref_point,
+                                      tangent_vec),
+                               inv_sqrt_ref_point)
+    riem_exp_from_id = group_exp(tangent_vec_at_id)
+
+    riem_exp = np.dot(sqrt_ref_point, np.dot(riem_exp_from_id, sqrt_ref_point))
+
+    return riem_exp
+
+
+def riemannian_log(ref_point, point):
+    """
+    Compute the Riemannian logarithm at point ref_point,
+    of point wrt the metric defined in
+    riemannian_inner_product.
+
+    This gives a tangent vector at point ref_point.
+    """
+    sqrt_ref_point = scipy.linalg.sqrtm(ref_point)
+    inv_sqrt_ref_point = np.linalg.inv(sqrt_ref_point)
+
+    point_near_id = np.dot(np.dot(inv_sqrt_ref_point,
+                                  point),
+                           inv_sqrt_ref_point)
+    riem_log_at_id = group_log(point_near_id)
+
+    riem_log = np.dot(np.dot(sqrt_ref_point, riem_log_at_id), sqrt_ref_point)
+
+    return riem_log
+
+
+def riemannian_mean(sym_matrices, n_max_iterations, epsilon=EPSILON):
+    """
+    Compute the Riemannian mean (Frechet mean) iterating 3 steps:
+    - Project all the matrices onto the tangent space using the riemannian log
+    - Calculate the tangent mean on the tangent space
+    - Shoot the tangent mean onto the manifold using the riemannian exp
+
+    Initialization with one of the matrices.
+    """
+
+    dists_between_iterates = []
+    n_sym_matrices = len(sym_matrices)
+
+    riem_mean = sym_matrices[0]
+
+    dim = len(riem_mean)
+
+    dist = math.inf
+    it = 0
+    while it < n_max_iterations and dist > epsilon:
```

Comment:
at some point when you want to deal with performance it might be interesting to profile this code and check where you're getting the biggest performance hit. Not for this code review but just out of curiosity :)